diff --git a/gdk/gdkwindow.c b/gdk/gdkwindow.c
index 0be4656..a643619 100644
--- a/gdk/gdkwindow.c
+++ b/gdk/gdkwindow.c
@@ -6546,6 +6546,8 @@ gdk_window_set_background_pattern (GdkWindow *window,
       GdkWindowImplClass *impl_class = GDK_WINDOW_IMPL_GET_CLASS (window->impl);
       impl_class->set_background (window, pattern);
     }
+  else
+    gdk_window_invalidate_rect_full (window, NULL, TRUE, CLEAR_BG_ALL);
 }

 /**
